pkg/policy/rule.go

```diff
@@ -66,21 +77,47 @@ func mergeL4IngressPort(ctx *SearchContext, endpoints []api.EndpointSelector, r
 		resMap[key] = CreateL4IngressFilter(endpoints, r, p, proto, ruleLabels)
 		return 1, nil
 	}
+
+	// If rule selects all endpoints and no L7 rules apply, then remove all
+	// L7-related information from the filter and return.
+	if filter.addEndpoints(endpoints) && r.NumRules() == 0 {
```

Comment:
I'm still not sure this is correct. You should do that only if the new rule selects a superset of the endpoints already selected by the filter.
I'm not sure that `filter.addEndpoints(endpoints)` means `the new rule selects a superset of the filter's endpoints`.